docs: added documentation for kubevirt on Talos (#9522)
siderolabs/talos: website/content/v1.8/advanced/install-kubevirt.md

---
title: "Install KubeVirt on Talos"
description: "This is a guide on how to get started with KubeVirt on Talos"
---

> KubeVirt allows you to run virtual machines on Kubernetes. It runs with QEMU and KVM to provide a seamless virtual machine experience and can be mixed with containerized workloads. This guide explains on how to install KubeVirt on Talos.

Author: I will fix your comments after the weekend. Nog problem!

> KubeVirt allows you to run virtual machines on Kubernetes. It runs with QEMU and KVM to provide a seamless virtual machine experience and can be mixed with containerized workloads. This guide explains on how to install KubeVirt on Talos.

Author: hi @rothgar,

I fixed the sentences by putting them on a new line.

## Prerequisites

For KubeVirt and Talos to work you have to enable certain configurations in the BIOS and configure Talos properly for it to work. 

### Enable virtualization in your BIOS

> On many new PCs and servers virtualization is enabled by default. Please consult your manufacturer on how to enable this in the BIOS. You can also run KubeVirt from within a virtual machine. For that to work you have to enable Nested Virtualization. This can also be done in the BIOS.

Reviewer: ```suggestion
On many new PCs and servers, virtualization is enabled by default. Please consult your manufacturer on how to enable this in the BIOS. You can also run KubeVirt from within a virtual machine. For that to work you have to enable Nested Virtualization. This can also be done in the BIOS.

You should verify with your cloud provider if they allow nested virtualization.
```

### Configure your network interface in bridge mode (optional)

> When you want to leverage [Multus](../kubernetes-guides/network/multus) to give your virtualm achines direct access to your nodes network, your bridge need to configured properly. This can be done by setting your network interface in bridge mode. You can look up the network interface name by using the following command:

Reviewer: ```suggestion
When you want to leverage [Multus](../kubernetes-guides/network/multus) to give your virtual machines direct access to your node network, your bridge needs to be configured properly. This can be done by setting your network interface in bridge mode. You can look up the network interface name by using the following command:
```

10.99.101.9   network     LinkStatus   cni0           9         ether      bridge   1e:5e:99:8f:1e:19                                 up           true
10.99.101.9   network     LinkStatus   dummy0         1         ether      dummy    62:1c:3e:d5:72:11                                 down         false
10.99.101.9   network     LinkStatus   eth0           5         ether               bc:24:11:a1:98:fc                         
```

In this case, this network interface is called `eth0`. Now you can configure your bridge properly. This can be done in the machine config of your node:

```yaml
machine:
      interfaces:
      - interface: br0
        addresses:
          - 10.99.101.9/24
        bridge:
          stp:
            enabled: true
          interfaces:
              - eth0 # This must be changed to your matching interface name
        routes:
            - network: 0.0.0.0/0 # The route's network (destination).
              gateway: 10.99.101.254 # The route's gateway (if empty, creates link scope route).
              metric: 1024 # The optional metric for the route.
``` 

### Install the `local-path-provisioner`

> When we are using KubeVirt, we are also installing the CDI (containerized data importer) operator. For this to work properly, we have to install the `local-path-provisioner`. This CNI kan be used to write scratch space when importing images with the CDI. 

Reviewer: ```suggestion
If you want to use the Containerized Data Importer (CDI) operator you will need to install a storage provider that offers a `StorageClass`. For testing you can install the `local-path-provisioner`. This CSI can be used to write scratch space when importing images with the CDI. 
```

> You can simply install the `local-path-provisioner` by following [this](../kubernetes-guides/configuration/local-storage) guide

Reviewer: ```suggestion
You can install the `local-path-provisioner` by following [this guide](../kubernetes-guides/configuration/local-storage).
```

### Configure storage

If you would like to use features such as `LiveMigration` shared storage is neccesary. You can either choose to install a CSI that connects to NFS or you can install Longhorn, for example. For more information on how to install Longhorn on Talos you can follow [this](https://longhorn.io/docs/1.7.2/advanced-resources/os-distro-specific/talos-linux-support/) link.

> To install the NFS-CSI driver, you can follow [This](https://github.com/kubernetes-csi/csi-driver-nfs/blob/master/docs/install-csi-driver-v4.9.0.md) guide. 

Reviewer: What's the difference between this link and the one above?

Author: The first one is the installation of Longhorn on Talos. The other one is the installation of the NFS-CSI driver. Which is neccesary if you want to connect to a NFS share.

After the installation of the NFS-CSI driver is done, you can create a storage class for the NFS CSI driver to work:

```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: nfs-csi
  annotations:
    storageclass.kubernetes.io/is-default-class: "true"
provisioner: nfs.csi.k8s.io
parameters:
  server: 10.99.102.253
  share: /mnt/data/nfs/kubernetes_csi
reclaimPolicy: Delete
volumeBindingMode: Immediate
mountOptions:
  - nfsvers=3
  - nolock
```

Note that this is just an example. Make sure to set the `nolock` option. If not, the nfs-csi storageclass won't work, because talos doesn't have a `rpc.statd` daemon running.

### Install `virtctl`

`virtctl` is needed for communication between the CLI and the KubeVirt api server. 

You can install the `virtctl` client directly by running:

```bash

> $ export VERSION=$(curl https://storage.googleapis.com/kubevirt-prow/release/kubevirt/kubevirt/stable.txt)

Reviewer: ```suggestion
export VERSION=$(curl https://storage.googleapis.com/kubevirt-prow/release/kubevirt/kubevirt/stable.txt)
```

> $ wget https://github.com/kubevirt/kubevirt/releases/download/${VERSION}/virtctl-${VERSION}-linux-amd64

Reviewer: ```suggestion
wget https://github.com/kubevirt/kubevirt/releases/download/${VERSION}/virtctl-${VERSION}-linux-amd64
```

```
Or you can use [krew](https://github.com/kubernetes-sigs/krew/#installation) to integrate it nicely in `kubectl`:

```bash

> $ kubectl krew install virt

Reviewer: ```suggestion
kubectl krew install virt
```

```

## Installing KubeVirt

After the neccesary preperations are done, you can now install KubeVirt. This can either be done through the [Operator Lifecycle Manager](https://olm.operatorframework.io/docs/getting-started/) or by just simply applying a YAML file. We will keep this simple and do the following:

```bash
# Point at latest release

> $ export RELEASE=$(curl https://storage.googleapis.com/kubevirt-prow/release/kubevirt/kubevirt/stable.txt)

Reviewer: ```suggestion
export RELEASE=$(curl https://storage.googleapis.com/kubevirt-prow/release/kubevirt/kubevirt/stable.txt)
```

# Deploy the KubeVirt operator

> $ kubectl apply -f https://github.com/kubevirt/kubevirt/releases/download/${RELEASE}/kubevirt-operator.yaml

Reviewer: ```suggestion
kubectl apply -f https://github.com/kubevirt/kubevirt/releases/download/${RELEASE}/kubevirt-operator.yaml
```

```

After the operator is installed, it is time to apply the Custom Resource (CR) for the operator to fully deploy KubeVirt. 

```yaml
---
apiVersion: kubevirt.io/v1
kind: KubeVirt
metadata:
  name: kubevirt
  namespace: kubevirt
spec:
  configuration:  
    developerConfiguration:
      featureGates: 
        - LiveMigration
        - NetworkBindingPlugins
    smbios:
      sku: "TalosCloud"
      version: "v0.1.0"
      manufacturer: "Talos Virtualization"
      product: "talosvm"
      family: "ccio"
  workloadUpdateStrategy:
    workloadUpdateMethods:

>     - LiveMigrate

Reviewer: ```suggestion
    - LiveMigrate # enable if you deployed Longhorn
```

In this example we enable:

- `LiveMigration` -- For live migration of virtual machines to other nodes
- `NetworkBindingPlugins` -- This is needed for Multus to work.

#### `smbios`

Here we configure a specific smbios configuration. This can be useful when you want to give your virtual machines a own sku, manufacturer name etc. 

#### `workloadUpdateStrategy`

If this is configured, virtual machines will be live migrated to other nodes when KubeVirt is updated.

## Installing CDI

The CDI (containerized data importer) is needed to import virtual disk images in your KubeVirt cluster. The CDI can do the following:

- Import images of type:
  - qcow2
  - raw
  - iso
- Import disks from either:
  - http/https
  - uploaded through virtctl
  - Container registry
  - Another PVC

You can either import these images by creating a DataVolume CR or by integrating this in your `VirtualMachine` CR. 

When applying either the `DataVolume` CR or the `VirtualMachine` CR with a `dataVolumeTemplates`, the CDI kicks in and will do the following:

- creates a PVC with the requirements from either the `DataVolume` or the `dataVolumeTemplates`

> - Start a pod

Reviewer: ```suggestion
- starts a pod
```

- writes temporary scratch space to local disk
- downloads the image
- extracts it to the temporary scratch space

> - Copies the image to the PVC

Reviewer: ```suggestion
- copies the image to the PVC
```

Installing the CDI is very simple:

```bash
# Point to latest release

> $ export TAG=$(curl -s -w %{redirect_url} \

Reviewer: ```suggestion
export TAG=$(curl -s -w %{redirect_url} \
```

https://github.com/kubevirt/containerized-data-importer/releases/latest)

> $ export VERSION=$(echo ${TAG##*/})

Reviewer: ```suggestion
export VERSION=$(echo ${TAG##*/})
```

# install operator

> $ kubectl create -f \

Reviewer: ```suggestion
kubectl create -f \
```

https://github.com/kubevirt/containerized-data-importer/releases/download/$VERSION/cdi-operator.yaml
```

After that, you can apply a CDI CR for the CDI operator to fully deploy CDI:

```yaml
apiVersion: cdi.kubevirt.io/v1beta1
kind: CDI
metadata:
  name: cdi
spec:
  config:
    scratchSpaceStorageClass: local-path
    podResourceRequirements:
      requests:
        cpu: "100m"
        memory: "60M"
      limits:
        cpu: "750m"
        memory: "2Gi"
```

This CR has some special settings that are needed for CDI to work properly:

### `scratchSpaceStorageClass`

This is the storage class that we installed earlier with the `local-path-provisioner`. This is needed for the CDI to write scratch space to local disk before importing the image

### `podResourceRequirements`

In many cases the default resource requests and limits are not sufficient for the importer pod to import the image. This will result in a crash of the importer pod.

After applying this yaml file, the CDI operator is ready.

> ## creating your first virtual machine

Reviewer: ```suggestion
## Creating your first virtual machine
```

```yaml
---
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: fedora-vm
spec:
  running: false
  template:
    metadata:
      labels:
        kubevirt.io/vm: fedora-vm
      annotations:
        kubevirt.io/allow-pod-bridge-network-live-migration: "true"
        
    spec:
      evictionStrategy: LiveMigrate
      domain:
        cpu:
          cores: 2
        resources:
          requests:
            memory: 4G
        devices:
          disks:
            - name: fedora-vm-pvc
              disk:
                bus: virtio
            - name: cloudinitdisk
              disk:
                bus: virtio
          interfaces:
          - name: podnet
            masquerade: {}    
      networks:
        - name: podnet
          pod: {}
      volumes:
        - name: fedora-vm-pvc
          persistentVolumeClaim:
            claimName: fedora-vm-pvc
        - name: cloudinitdisk
          cloudInitNoCloud:
            networkData: |
              network:
                version: 1
                config:
                  - type: physical

>                     name: eth0

Reviewer: Does this use the bridge adapter?

Author: No, this uses the pod network. Should i also add an example for bridge network configuration?

Reviewer: Because we mention setting up multus and using a bridge we should probably include an example or at least a commented out section in the yaml to show how they'd use it instead of CNI.

Author: > Because we mention setting up multus and using a bridge we should probably include an example or at least a commented out section in the yaml to show how they'd use it instead of CNI.

I will add an example tonight. And will fix the other issues. Thanks for replying :)

Author: @rothgar i added an Multus example to the documentation.
